### ISSUE TITLE
Unexpected PossiblyUndefinedVariable with empty(mixed[offset])

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#372]
Apparently there's a problem with `empty()` calls on the results of array access calls on mixed variables. I can't quite point the finger on it, but I don't think _any_ of the following should report any errors, but some do (see inline/dockblock comments):
```php
<?php
// results in two PossiblyUndefinedVariable in if conditions
function a(array $in): void
{
    $r = [];
    foreach ($in as $entry) {
        if (!empty($entry['a'])) {
            $r[] = [];
        }
        if (empty($entry['a'])) {
            $r[] = [];
        }
    }
}

// results in one PossiblyUndefinedVariable (second if condition)
function b(array $in): void
{
    $i = 0;
    foreach ($in as $entry) {
        if (!empty($entry['a'])) {
            $i--;
        }
        if (empty($entry['a'])) {
            $i++;
        }
    }
}

/**
 * results in one PossiblyUndefinedVariable (second if)
 */
function c(array $in): void
{
    foreach ($in as $entry) {
        if (!empty($entry['a'])) {}
    }
    foreach ($in as $entry) {
        if (empty($entry['a'])) {}
    }
}

/**
 * results in NO problems
 */
function d(array $in): void
{
    foreach ($in as $entry) {
        if (empty($entry['a'])) {}
    }
}

/**
 * NO problems either
 */
function e(array $in): void
{
    foreach ($in as $entry) {
        if (!empty($entry['a'])) {}
    }
}

/**
 * Results in one PossiblyUndefinedVariable (second if)
 */
function f(array $in): void
{
    $entry = reset($in);
    if (!empty($entry['a'])) {}
    if (empty($entry['a'])) {}
}

/**
 * Results in one PossiblyUndefinedVariable (second if)
 * @param mixed $in
 */
function g($in): void
{
    if (!empty($in['a'])) {}
    if (empty($in['a'])) {}
}

/**
 * NO problems reported here
 * @param mixed $in
 */
function h($in): void
{
    if (isset($in['a']) && $in['a']) {}
    if (!isset($in['a']) || !$in['a']) {}
}
```

**Comment by muglug** [comment]
Yes, `!empty` should work like `isset` and `!falsy` combined, but currently is just `!falsy`. Will fix

**Comment by weirdan** [comment]
Why there's a difference in number of problems reported in `a()` and `b()`? To me this suggests there could be some other problem.

**Comment by muglug** [comment]
That's a slight difference with how loops are evaluated. Here we get the number of times Psalm thinks it has to analyse the loop in order to fully understand the series of assignments: https://github.com/vimeo/psalm/blob/e2b1b24dbfa4a4fafd57de5558d289c28f14bf63/src/Psalm/Checker/Statements/Block/LoopChecker.php#L79

In the `function a`, `$r` is just assigned a constant value. In `function b`, `$i` is assigned to a value based on a variable - the AssignmentMapper treats it as `$i = $i + 1;` (though it turns out that the `++` and `--` operators never change the type, so this can be optimised away)

With that in mind, it runs the analysis twice, and only emits the errors still present the second time around. Having fixed the bug, there are no errors either time now.

------

Fun fact:
```php
$a = "hello";
$a++;
echo $a; // hellp
```

Whereas if you replace `$a++` with `$a = $a + 1` you get a warning

**Comment by weirdan** [comment]
> In the function a, $r is just assigned a constant value.

Uhm, no, `$r[] = []; ` pushes empty array as the last element of `$r`. So it should behave similarly to `$i++`:
```php
// if
$i++
// is like
$i = $i + 1;
// then
$r[] = [];
// should be like 
$r = array_merge($r, [[]]); // or array_push($r, []);
```
```
$ php -r '$a = []; for ($i = 0; $i<3; $i++) { $a[] = rand(); } var_dump($a);'
Command line code:1:
array(3) {
  [0] =>
  int(229307675)
  [1] =>
  int(1600905179)
  [2] =>
  int(342487633)
}
```
Variable value changes, but type doesn't.

**Comment by weirdan** [comment]
You seem to imply that for `a()` (with array push) Psalm does one pass, and for `b()` (with increment) it does two passes, but it reported _two_ "problems" for `a()` and only complained _once_ for `b()`

**Comment by muglug** [comment]
Understood. But the heuristic to figure out how many times to analyse the loop doesn’t analyse type changes - just assignments from one var to another. And in treating `$i++` as `$i = $i + 1`, it’s treating it the same as `$i = [$i]` (and any other assignment that did change the type of `$i`.

**Comment by muglug** [comment]
Oh I'm an idiot. Should have looked at the actual code.

You're totally right – we only look at assignments, not `PostInc` or other `$a++`-like operations. So that's why the `$r[] = []` had more checks than `$i++`, and hence fewer loops
